dandi/upload.py

```diff
@@ -573,3 +587,272 @@ def upload_agg(*ignored):
+        """
+
+        Parameters
+        ----------
+        path: Path
+          Non Pure (OS specific) Path
+        relpath:
+          For location on server.  Will be cast to PurePosixPath
+
+        Yields
+        ------
+        dict
+          Records for pyout
+        """
+        # Ensure consistent types
+        path = Path(path)
+        relpath = PurePosixPath(relpath)
+        try:
+            try:
+                path_stat = path.stat()
+                yield {"size": path_stat.st_size}
+            except FileNotFoundError:
+                yield skip_file("ERROR: File not found")
+                return
+            except Exception as exc:
+                # without limiting [:50] it might cause some pyout indigestion
+                yield skip_file("ERROR: %s" % str(exc)[:50])
+                return
+
+            file_metadata_ = {
+                "uploaded_size": path_stat.st_size,
+                "uploaded_mtime": ensure_strtime(path_stat.st_mtime),
+                # "uploaded_date": None,  # to be filled out upon upload completion
+            }
+
+            #
+            # Validate first, so we do not bother server at all if not kosher
+            #
+            # TODO: enable back validation of dandiset.yaml
+            if path.name != dandiset_metadata_file and validation != "skip":
+                yield {"status": "validating"}
+                validation_errors = validate_file(path)
+                yield {"errors": len(validation_errors)}
+                # TODO: split for dandi, pynwb errors
+                if validation_errors:
+                    if validation == "require":
+                        yield skip_file("failed validation")
+                        return
+                else:
+                    yield {"status": "validated"}
+            else:
+                # yielding empty causes pyout to get stuck or crash
+                # https://github.com/pyout/pyout/issues/91
+                # yield {"errors": '',}
+                pass
+
+            #
+            # Special handling for dandiset.yaml
+            # Yarik hates it but that is life for now. TODO
+            #
+            if path.name == dandiset_metadata_file:
+                # TODO This is a temporary measure to avoid breaking web UI
+                # dandiset metadata schema assumptions.  All edits should happen
+                # online.
+                yield skip_file("should be edited online")
+                return
+
+            #
+            # Extract metadata - delayed since takes time, but is done before
+            # actual upload, so we could skip if this fails
+            #
+            # Extract metadata before actual upload and skip if fails
+            # TODO: allow for for non-nwb files to skip this step
+            # ad-hoc for dandiset.yaml for now
+            yield {"status": "extracting metadata"}
+            try:
+                metadata = get_metadata(path)
```

Comment:
Should just `metadata2asset()` be used, or should `nwb2asset()` be used instead so that `contentSize`, `encodingFormat`, etc. are automatically populated?

Reply:
rright -- use `nwb2asset`! but it should be RFed so we can get similar record for non-nwb files/assets as well